File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        int[] answer;
        List<Integer> answerList = new ArrayList<>();
        String[] arr = msg.split("");
        int idx = 1;
        Map<String, Integer> dic = new HashMap<>();
        Deque<String> q = new ArrayDeque<>();
        String word = "";
        
        for (String alpha:"ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("")) {
            dic.put(alpha, idx++);
        }
        
        for(String alpha:msg.split("")) {
            q.addLast(alpha);
        }
        
        while (!q.isEmpty()) {
            word += q.removeFirst();
            while (!q.isEmpty() && dic.get(word) != null) {
                word += q.removeFirst();
            }
            System.out.println(word);
            if (dic.get(word) != null) {
                answerList.add(dic.get(word));
            } else {
                answerList.add(dic.get(word.substring(0, word.length()-1)));
                dic.put(word, idx++);
                q.addFirst(word.substring(word.length()-1));
                word = "";
            }
        }
        
        answer = new int[answerList.size()];
        for (int i=0;i<answer.length;i++) {
            answer[i] = answerList.get(i);
        }
        return answer;
    }
}

File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.py
from collections import deque

def solution(msg):
    idx = 1
    dic = {}
    
    for _ in range(26):
        dic[chr(ord("A") + idx - 1)] = idx
        idx += 1

    q = deque(list(msg))

    answer = []
    while q:
        temp = q.popleft()
        
        while q and temp in dic:
            temp += q.popleft()

        if temp in dic:
            answer.append(dic[temp])
        else:
            answer.append(dic[temp[:-1]])
            dic[temp] = idx
            idx += 1
            q.appendleft(temp[-1])

    return answer